src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs

```diff
@@ -5179,6 +5179,16 @@ protected internal override SyntaxNode GetTopmostNodeForDiagnosticAnalysis(ISymb
                         return fieldDecl;
                     }
 
+                    break;
+
+                case SymbolKind.Method:
+                    if (declaringSyntax is ArrowExpressionClauseSyntax &&
+                        declaringSyntax.Parent is PropertyDeclarationSyntax or IndexerDeclarationSyntax)
+                    {
+                        Debug.Assert(((IMethodSymbol)symbol).MethodKind == MethodKind.PropertyGet);
```

Comment:
1. what happens for other accessors (like setters)?
2. PropertyDeclarationSyntax/IndexerDeclarationSyntax share a common BasePropertyDeclarationSyntax, can you use that?
3. why only for property/indexer.  Is there a concern about add/remove for events?
4. why only for property/indexer?  What about methods/constructors/destructors/operarators?

If hte answer is: it works for all those other cases, can we add tests?

Reply:
oh, i see why this isn't for events. this is not for the accessor case, but when you have `int P => ...`  still unsure about all the other method/constructor/destructor/operator cases.

Reply:
I also don't understand this.

Reply:
so the only time when you'd have a `=>` parented by a BaseProperty is if you have `int Foo => ...` or `int this[...] => ...`.  In either case, the ArrayExpr corresponds to the GetAccessor and nothing else (that's how i understand the code).

Reply:
> so the only time when you'd have a => parented by a BaseProperty is if you have int Foo => ... or int this[...] => .... In either case, the ArrayExpr corresponds to the GetAccessor and nothing else (that's how i understand the code).

That is correct.

However, I have put some more thought into this, and I think we want to allow analyzers to report diagnostics anywhere within the Property/Indexer declaration sub-tree from operation/node callbacks inside accessory body, regardless of it being an expression body or a block body. For example, if an analyzer that gets a callback for literal `0` inside accessor body and reports a diagnostic on the identifier `P`, it would be weird if the diagnostic is classified as a local diagnostic for expression body case, but not for block body case. I believe we want both the cases to be treated uniformly.

I am going to update the logic, add more tests for expression body and block body for all method kinds: accessors, ctors, dtors, operators, ordinary methods, etc. that validate all of them have identical behavior.

Reply:
@mavasani that makes a lot of sense to me :)

Reply:
I also figured out that the current implementation in the PR will change the Node/Operation passed into CodeBlockStart/OperationBlockStart callbacks. That would be a breaking change, which we definitely want to avoid. I am going to change the implementation a bit to avoid this and instead just adjust the scope for local/non-local diagnostic classification. I'll also add this validation in the added unit tests.

Reply:
I have pushed an updated implementation.